src/nti/schema/schema.py

```diff
@@ -131,12 +155,72 @@ def __init__(self, **kw):
             if k not in schema:
                 raise TypeError('non schema keyword argument: %s' % k)
             setattr(self, k, v)
-        # provide default values for schema fields not set
-        for f, fields in schema.items():
-            if getattr(self, f, _marker) is _marker:
+
+        # provide default values for schema fields not set.
+        # In bench_schemaconfigured.py, if the fields are FieldProperty objects found in the
+        # type, checking for whether they are set or not took 96% of the total time.
+        # We can be much faster (33us -> 9.1us) if we special case this, without hurting
+        # the non-FieldProperty case too much.
+        if self.SC_OPTIMIZE_FIELD_PROPERTY:
+            schema = self.__elide_fieldproperty(schema)
+
+        for field_name, schema_field in schema.items():
+            if field_name in kw:
+                continue
+            # TODO: I think we could do better by first checking
+            # to see if field_name is in vars(type(self))?
+            if getattr(self, field_name, _marker) is _marker:
                 # The point of this is to avoid hiding exceptions (which the builtin
                 # hasattr() does on Python 2)
-                setattr(self, f, fields.default)
+                setattr(self, field_name, schema_field.default)
+
+    __FP_KEY = '__SchemaConfigured_elide_fieldproperty'
+
+    @classmethod
+    def __elide_fieldproperty(cls, schema):
+        try:
+            matches = cls.__dict__[cls.__FP_KEY]
+        except KeyError:
+            matches = cls.__find_FieldProperty_that_match_schema(schema)
+            setattr(cls, cls.__FP_KEY, matches)
+
+        return {k: v for k, v in schema.items() if k not in matches}
+
+
+    @classmethod
+    def __find_FieldProperty_that_match_schema(cls, schema_dict):
+        result = set()
+        for field_name, schema_field in schema_dict.items():
+            try:
+                # If these are descriptors, this runs code. We don't look in the
+                # type's __dict__ because we would need to manually walk up the mro().
+                kind_value = getattr(cls, field_name)
+            except AttributeError:
+                continue
+
+            # pylint:disable=protected-access
+            if isinstance(kind_value, FieldProperty) \
+               and kind_value._FieldProperty__field == schema_field:
+                # These are data-descriptors, with both __get__ and
+                # __set__, so they're in full control. They automatically return
+                # the default value of the field (that they have), so we don't
+                # need to copy it down from our schema field.
+                result.add(field_name)
+        return result
```

Comment:
So we are avoiding setting the default value on the instance for attributes defined by FieldProperty descriptors which themselves handle returning a default value when not set. We've combined that with caching the set of attributes that aren't defined by FieldProperty descriptors so the next creation of this class doesn't have to do that check. Did I follow all that correctly?

I've always found it odd that defaults were set at creation time rather than simply provided at read time. Especially when combined with persistent objects. Defaults get baked in at creation time. If I'm following all this, this change makes that behaviour inconsistent depending on if the attribute is a FieldProperty or not. I could imagine that discrepancy being surprising so it's probably worth an explicit mention somewhere. FWIW I prefer the new behaviour for FieldProperty attributes. If you need to bake in the defaults at creation time you can just explicitly set it.

Reply:
> Did I follow all that correctly?

Pretty much. The key observation is that if the schema and the FieldProperty are using the same field, with the same default, nothing should change. This is even more so given that...

> I've always found it odd that defaults were set at creation time rather than simply provided at read time. 

...they're not, not for `FieldProperty` objects. `getattr(self, field_name, _marker)` is going to wind up in `FieldProperty.__get__`, which will bind the schema field to the object and *return* (but never actually store) the bound field's default. That doesn't raise an `AttributeError`, (assuming the field actually has a default, which it should) so `getattr(.., _marker) is _marker` is always going to be False. Thus, the `setattr` is never executed. But, the work for binding the field (and any computation it does in `.default`) is thrown away. (Profiling showed the biggest expense here was in binding the field.)

>  If I'm following all this, this change makes that behaviour inconsistent depending on if the attribute is a FieldProperty or not….

I don't think so. But what is now inconsistent is the handling of `FieldPropertyStoredThroughField`, though that should just be a speed discrepancy. I thought those were probably rare enough not to care about.

Reply:
That makes sense. Thanks.